File: python/src/ev3/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundocentralita/centralita_resuelto.py
"""Centralita CES — espejo Kotlin."""
from abc import ABC, abstractmethod


class Llamada(ABC):
    def __init__(self, n_origen: str, n_destino: str, duracion: int):
        self.n_origen = n_origen
        self.n_destino = n_destino
        self.duracion = duracion
        self.coste = self.calcular_coste()

    @abstractmethod
    def calcular_coste(self) -> float:
        ...

    def mostrar_datos(self) -> None:
        print(f"Origen = {self.n_origen}")
        print(f"Destino = {self.n_destino}")
        print(f"Duracion = {self.duracion}")
        print(f"Coste = {self.coste}")


class LlamadaLocal(Llamada):
    def calcular_coste(self) -> float:
        return 0.0

    def mostrar_datos(self) -> None:
        print("Tipo = Local")
        super().mostrar_datos()


class LlamadaProvincial(Llamada):
    def calcular_coste(self) -> float:
        return self.duracion * 0.15

    def mostrar_datos(self) -> None:
        print("Tipo = Provincial")
        super().mostrar_datos()


class LlamadaNacional(Llamada):
    def __init__(self, n_origen: str, n_destino: str, duracion: int, franja: int):
        self.franja = franja
        super().__init__(n_origen, n_destino, duracion)

    def calcular_coste(self) -> float:
        precios = {1: 0.20, 2: 0.25, 3: 0.30}
        return self.duracion * precios.get(self.franja, 0.20)

    def mostrar_datos(self) -> None:
        print("Tipo = Nacional")
        print(f"Franja = {self.franja}")
        super().mostrar_datos()


class Centralita:
    def __init__(self):
        self.llamadas: list[Llamada] = []

    def registrar(self, llamada: Llamada) -> None:
        self.llamadas.append(llamada)
        print("Llamada registrada")
        llamada.mostrar_datos()

    def mostrar(self) -> None:
        if not self.llamadas:
            print("No hay llamadas registradas")
            return
        for l in self.llamadas:
            l.mostrar_datos()

    def costes_totales(self) -> None:
        print(f"Coste total = {sum(l.coste for l in self.llamadas)}")


def main() -> None:
    c = Centralita()
    opcion = 0
    while opcion != 6:
        print("1 Local 2 Provincial 3 Nacional 4 Mostrar 5 Costes 6 Salir")
        opcion = int(input("Opcion: "))
        if opcion == 1:
            c.registrar(LlamadaLocal(input("Origen: "), input("Destino: "), int(input("Duracion: "))))
        elif opcion == 2:
            c.registrar(LlamadaProvincial(input("Origen: "), input("Destino: "), int(input("Duracion: "))))
        elif opcion == 3:
            c.registrar(LlamadaNacional(
                input("Origen: "), input("Destino: "), int(input("Duracion: ")), int(input("Franja: "))
            ))
        elif opcion == 4:
            c.mostrar()
        elif opcion == 5:
            c.costes_totales()
        elif opcion == 6:
            print("Saliendo")


if __name__ == "__main__":
    main()

File: csharp/src/ev3/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundocentralita/CentralitaResuelto.cs
// Centralita CES — espejo Kotlin
using System;
using System.Collections.Generic;

abstract class Llamada {
    public string NOrigen, NDestino; public int Duracion; public double Coste;
    public Llamada(string o, string d, int dur) { NOrigen=o; NDestino=d; Duracion=dur; Coste=CalcularCoste(); }
    public abstract double CalcularCoste();
    public virtual void MostrarDatos() {
        Console.WriteLine($"Origen = {NOrigen}"); Console.WriteLine($"Destino = {NDestino}");
        Console.WriteLine($"Duracion = {Duracion}"); Console.WriteLine($"Coste = {Coste}");
    }
}
class LlamadaLocal : Llamada {
    public LlamadaLocal(string o,string d,int dur):base(o,d,dur){}
    public override double CalcularCoste() => 0;
    public override void MostrarDatos(){ Console.WriteLine("Tipo = Local"); base.MostrarDatos(); }
}
class LlamadaProvincial : Llamada {
    public LlamadaProvincial(string o,string d,int dur):base(o,d,dur){}
    public override double CalcularCoste() => Duracion * 0.15;
    public override void MostrarDatos(){ Console.WriteLine("Tipo = Provincial"); base.MostrarDatos(); }
}
class LlamadaNacional : Llamada {
    public int Franja;
    public LlamadaNacional(string o,string d,int dur,int f):base(o,d,dur){ Franja=f; Coste=CalcularCoste(); }
    public override double CalcularCoste() => Duracion * (Franja==2?0.25:Franja==3?0.30:0.20);
    public override void MostrarDatos(){ Console.WriteLine("Tipo = Nacional"); Console.WriteLine($"Franja = {Franja}"); base.MostrarDatos(); }
}
class Centralita {
    public List<Llamada> Llamadas = new();
    public void Registrar(Llamada l){ Llamadas.Add(l); Console.WriteLine("Llamada registrada"); l.MostrarDatos(); }
    public void Mostrar(){ foreach(var l in Llamadas) l.MostrarDatos(); }
    public void Costes(){ double t=0; foreach(var l in Llamadas) t+=l.Coste; Console.WriteLine($"Coste total = {t}"); }
}
class CentralitaResuelto {
    static void Main() {
        var c = new Centralita(); int op;
        do {
            Console.WriteLine("1 Local 2 Provincial 3 Nacional 4 Mostrar 5 Costes 6 Salir");
            op = int.Parse(Console.ReadLine()!);
            if (op==1) c.Registrar(new LlamadaLocal(Console.ReadLine()!, Console.ReadLine()!, int.Parse(Console.ReadLine()!)));
            else if (op==2) c.Registrar(new LlamadaProvincial(Console.ReadLine()!, Console.ReadLine()!, int.Parse(Console.ReadLine()!)));
            else if (op==3) c.Registrar(new LlamadaNacional(Console.ReadLine()!, Console.ReadLine()!, int.Parse(Console.ReadLine()!), int.Parse(Console.ReadLine()!)));
            else if (op==4) c.Mostrar();
            else if (op==5) c.Costes();
        } while(op!=6);
    }
}
